lib: os: heap: add callbacks for allocation and free

dcpleung opened the pull request:
This adds the ability to be notified when heap allocation and free operations are performed.

For example, this will allow power managing memory chips if we know which chips are actively being used.

andyross reviewed (approved):
Seems very reasonable to me.   One thing you might consider would be adding an explicit warning about synchronization.  The sys_heap layer isn't synchronized at all, which means that it's likely any user will be doing the work inside a lock of its own, which means that any callbacks will be implicitly under the same lock environment (and thus might deadlock if you try to call back into the same heap!).

Obviously none of that is a problem to be solved by the code here, but it strikes me as a likely footgun for future users.

MaureenHelm commented on lib/os/heap.c:
```diff
@@ -8,6 +8,19 @@
 #include <string.h>
 #include "heap.h"
 
+static inline void do_callback(struct z_heap *h,
+			       enum sys_heap_cb_ops op,
+			       void *mem, size_t bytes)
+{
+#ifdef CONFIG_SYS_HEAP_ALLOC_FREE_CALLBACKS
+	struct sys_heap *sh = CONTAINER_OF(h, struct sys_heap, heap);
+
+	if (sh->alloc_free_cb != NULL) {
+		sh->alloc_free_cb(op, mem, bytes);
+	}
+#endif
+}
```
I don't see this being used anywhere. Remove?

dcpleung replied:
Removed.

MaureenHelm commented on tests/lib/heap/prj.conf:
```diff
@@ -1,3 +1,4 @@
 CONFIG_ZTEST=y
 CONFIG_SYS_HEAP_VALIDATE=y
 CONFIG_SYS_HEAP_RUNTIME_STATS=y
+CONFIG_SYS_HEAP_ALLOC_FREE_CALLBACKS=y
```
These changes belong in commit `tests: lib/heap: add test for alloc/free callbacks`

MaureenHelm commented on tests/lib/heap_align/prj.conf:
```diff
@@ -1,3 +1,4 @@
 CONFIG_ZTEST=y
 CONFIG_SYS_HEAP_VALIDATE=y
 CONFIG_SYS_HEAP_RUNTIME_STATS=y
+CONFIG_SYS_HEAP_ALLOC_FREE_CALLBACKS=y
```
These changes belong in commit `tests: lib/heap: add test for alloc/free callbacks`

Damian-Nordic commented:
@dcpleung it seems I'm adding a similar functionality for the libc heap in #40117. @nashif suggests that we unify the solutions which would probably mean that per-heap instance callbacks that you're adding would need to become global (since there's no a notion of instance in the case of the libc heap). What is your opinion on that?

dcpleung commented:
I think per-instance callback is useful as you only listen to what you want to know. For example, I/O using dynamic memory for buffers will generate quite a number of events, where the application may not want to know.

cfriedt commented:
Just adding DNM until things are sorted out

KangJianX commented:
Verified this PR and it works.

dcpleung committed:
sys: heap_listener: resize_cb to also take heap ID

Add a parameter to the resize callback to also take the heap ID.
This allows a single callback to be used for multiple heaps if
so desired.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
sys: heap_listener: extend to cover more events

This extends the heap_listener to cover more events,
specifically, allocation, free and realloc.

Note that typedef are used so the callback can be
documented.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
sys: heap: add support for heap listener

This adds notification in the heap code to emit events for
heap listeners.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung committed:
tests: lib/heap: add test for heap listeners

This adds test to make heap listeners are working correctly.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented:
Updated to be based on the new heap listener APIs.

nashif commented:
@Damian-Nordic  please review

npitre commented:
A few comments:

- You should use `HEAP_ID_FROM_POINTER()` instead of `POINTER_TO_UINT()`.
  Same result but more coherent and self documenting. There should be
  a dummy version for when `CONFIG_HEAP_LISTENER` is unset (may simply
  return NULL).

- In `sys_heap_alloc()` please move the `CONFIG_SYS_HEAP_RUNTIME_STATS`
  down so the accounting and notifying stuff are kept close together.

- The notification in `sys_heap_aligned_realloc()` is wrong. Numbers won't
  balance.

  - You must not use `bytes` directly. That's not the amount of memory
    that will be freed later.

  - In the "Shrink in place, split off and free unused suffix" case, you
    must notify_free with `chunksz_to_bytes(h, c)` before `split_chunks()`
    and then notifi_alloc with `chunksz_to_bytes(h, chunk_size(h, c))`
    before the `return ptr`.

  - Same idea in the "Expand: split the right chunk and append" case:
    the notifi_free must be done before `merge_chunks()` and not use
    `bytes` directly.

dcpleung commented:
> * You should use `HEAP_ID_FROM_POINTER()` [...]
> * In `sys_heap_alloc()` please move the `CONFIG_SYS_HEAP_RUNTIME_STATS`
>   down so the accounting and notifying stuff are kept close together.

Fixed.

> * The notification in `sys_heap_aligned_realloc()` is wrong. Numbers won't
>   balance.
>   * You must not use `bytes` directly. That's not the amount of memory
>     that will be freed later.
>   * In the "Shrink in place, split off and free unused suffix" case, you
>     must notify_free with `chunksz_to_bytes(h, c)` before `split_chunks()`
>     and then notifi_alloc with `chunksz_to_bytes(h, chunk_size(h, c))`
>     before the `return ptr`.
>   * Same idea in the "Expand: split the right chunk and append" case:
>     the notifi_free must be done before `merge_chunks()` and not use
>     `bytes` directly.

Fixed for the "shink in place" case.

For the "expand" case, I moved the number of bytes to be freed calculation at the beginning of the if-block. Would this work? Or do I need to do the calculation just before `merge_chunks()`?

npitre commented:
> For the "expand" case, I moved the number of bytes to be freed calculation at the beginning of the if-block. Would this work? Or do I need to do the calculation just before `merge_chunks()`?

As long as it is done before c is modified with `merge_chunks()` it is 
fine. And the alloc size should be computed after.

npitre reviewed (approved):
The added ifdef's aren't that great but that's nitpicking

dcpleung commented:
Hm... lots of networking tests timed out.

dcpleung committed:
lib: os: CONFIG_HEAP_LISTENER to be hidden

This changes CONFIG_HEAP_LISTENER to be a hidden kconfig so that
the actual heap implementation can select it to enable
notifications. Each heap implementations will have their own
kconfigs to enable heap listener functionality so that app
can be built to only listen to certain heap implementations.

Signed-off-by: Daniel Leung <daniel.leung@intel.com>

dcpleung commented on lib/os/heap.c:
```diff
@@ -263,10 +269,17 @@ void *sys_heap_alloc(struct sys_heap *heap, size_t bytes)
 #endif
-	return chunk_mem(h, c);
+
+	heap_listener_notify_alloc(HEAP_ID_FROM_POINTER(heap), mem,
```
Done

npitre commented:
```
/*
 * Notify allocation first in those cases where reference
 * counting of allocated memory is needed (for example,
 * whether to power down memory banks). If memory free
 * is notified first, the ref count might go to zero
 * resulting in memory banks being powered down and
 * content wiped.
 */
```

If that's really the case then this cannot work as is anyway. If content is
lost then the heap metadata will be lost as well and the heap will no longer be
usable.

dcpleung commented:
> If that's really the case then this cannot work as is anyway. If content is lost then the heap metadata will be lost as well and the heap will no longer be usable.

So the metadata is stored throughout the heap buffer? If so, I will remove the comment to avoid giving out incorrect information.

npitre commented:
> So the metadata is stored throughout the heap buffer? If so, I will remove
> the comment to avoid giving out incorrect information.

Well, when it is all free, the metadata is concentrated at the beginning with a
small piece at the end. But the outcome is the same. If memory content is lost,
the heap will have to be reinitialized afresh when memory is available again.

Furthermore, if the notifier is triggering a memory shutdown, what does trigger
a memory power-up? There are many synchronization implications that are not
addressable with this mechanism.

dcpleung commented:
> Well, when it is all free, the metadata is concentrated at the beginning with a small piece at the end. But the outcome is the same. If memory content is lost, the heap will have to be reinitialized afresh when memory is available again.
> 
> Furthermore, if the notifier is triggering a memory shutdown, what does trigger a memory power-up? There are many synchronization implications that are not addressable with this mechanism.

Removed the comment.

If power cycling is needed, https://github.com/zephyrproject-rtos/zephyr/pull/40656 might work since the bookkeeping is done outside the buffer. However, that can't do allocation of arbitrary size, and the allocation is not continuous.

npitre commented:
@dcpleung: Please fix the following:

```
#ifdef CONFIG_SYS_HEAP_LISTENER
                size_t bytes_freed = chunksz_to_bytes(h, c);
#endif
```

The above will produce garbage.
You must wrap `c` with `chunk_size()`.

dcpleung commented:
> @dcpleung: Please fix the following:
> 
> ```
> #ifdef CONFIG_SYS_HEAP_LISTENER
>                 size_t bytes_freed = chunksz_to_bytes(h, c);
> #endif
> ```
> 
> The above will produce garbage. You must wrap `c` with `chunk_size()`.

Fix is at https://github.com/zephyrproject-rtos/zephyr/pull/41704
